packages/server-wallet/src/wallet/wallet.ts

```diff
@@ -66,18 +64,15 @@ export class Wallet {
    * @param objectiveMessages The collection of outgoing messages related to the objective.
    * @returns A promise that resolves when all the objectives are completed
    */
-  private async ensureObjectives(
-    objectives: WalletObjective[],
+  private async ensureObjective(
```

Comment:
I'm not sure about this change. It would trigger a flurry of network messaging when a client calls `createChannels(Array(10).fill(withParams))`.

You could instead change the return value of `ensureObjectives` to return `Promise<(ObjectiveSuccess | ObjectiveFailure)[]>`.

That has the downside of `createChannels` needing to do something like
```
const allDonePromise = this.ensureObjectives([...]);
const donePromiseForObjective5 = allResults.then(results => results[5])
```

to construct the corresponding promise for objective 5.

Is there a more elegant solution?

Reply:
Good point!

Let me take a look into this.

Reply:
So this actually turned out to be tricker than I thought. When switching to `ensureObjectives` it becomes non-trivial to setup the promises for each individual objective.

I've done this in f64a972 but would appreciate a look to see if there is nicer way of doing it.

Reply:
Another option is to push the complexity into the message service. It can collect and batch outgoing messages, for instance.

Reply:
> Another option is to push the complexity into the message service. It can collect and batch outgoing messages, for instance.

Hmm, that's a good point. I think I'd lean towards the `wallet` owning most of that complexity because
- The wallet better context of what should and should not be batched together than a message service
- The message service implementation should probably be as small/simple as possible so it's easy to implement for various communication protocols.

I could see the message service implementing batching on top of that though.

Reply:
Maybe we probably shouldn't rely on the user supplied message service to do batching. But, different channels may have differing capacity to batch messages, so it's not crazy. (Maybe my HHTP endpoint accepts up to 1Mb of data per request, but yours accepts only 256Kb.)

Reply:
Thinking about this further I think I agree with you. It seems simpler for the MessageService to implement batching if needed.